src/journal/src/Journal/Types.hs

```diff
@@ -19,20 +21,258 @@ module Journal.Types
   , getMaxByteSize
   , readFileCount
   , bumpFileCount
-  , module Journal.Types.AtomicCounter)
+  , module Journal.Types.AtomicCounter
+  , packTail
+  , termId
+  , termOffset
+  , align
+  )
   where
 
 import Control.Concurrent.STM
 import Control.Concurrent.STM (TVar)
 import Data.ByteString (ByteString)
 import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.Bits
+import Data.Vector (Vector)
 import Data.Word (Word32, Word64, Word8)
 import Foreign.Ptr (Ptr, plusPtr)
+import Foreign.Storable (sizeOf)
 
+import Journal.Internal.ByteBuffer
 import Journal.Types.AtomicCounter
 
 ------------------------------------------------------------------------
 
+-- XXX: `casIntArray` only works on `Int`, so we can't use `Int32` and `Int64`
+-- yet. According to
+-- https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/ghc-prim/changelog.md#080-edit-as-necessary
+-- casInt{32,64}Array should be part of ghc-prim 0.8.0, but the uploaded hackage
+-- package's changelog says it isn't part of that release, nor do the haddocks
+-- include it... Once this has been fixed we can remove the following type
+-- aliases:
+type Int64 = Int
+type Int32 = Int
+
+pARTITION_COUNT :: Int
+pARTITION_COUNT = 3
+
+data Journal' = Journal'
+  { jTermBuffers   :: {-# UNPACK #-} !(Vector ByteBuffer)
+  , jMetadata      :: {-# UNPACK #-} !ByteBuffer
+  , jPositionLimit :: {-# UNPACK #-} !AtomicCounter -- ???
+  }
+
+data JMetadata = JMetadata
+  { mdRawTail0      :: Int64
+  , mdRawTail1      :: Int64
+  , mdRawTail2      :: Int64
+  , mdActiveCount   :: Int32
+  -- padding
+  , mdInitialTermId :: Int32
+  -- mdDefaultFrameHeaderLength :: Int32?
+  -- mdMTULength :: Int32, only needed if we want to fragment large messages...
+  , mdTermLength    :: Int32
+  , mdPageSize      :: Int32
+  -- padding
+  -- , mdDefaultFrameHeader :: Bytestring???
+  }
+
+tERM_TAIL_COUNTERS_OFFSET :: Int
+tERM_TAIL_COUNTERS_OFFSET = 0
+
+lOG_ACTIVE_TERM_COUNT_OFFSET :: Int
+lOG_ACTIVE_TERM_COUNT_OFFSET = tERM_TAIL_COUNTERS_OFFSET +
+  sizeOf (8 :: Int64) * pARTITION_COUNT
+
+lOG_INITIAL_TERM_ID_OFFSET :: Int
+lOG_INITIAL_TERM_ID_OFFSET = lOG_ACTIVE_TERM_COUNT_OFFSET +
+  sizeOf (4 :: Int32)
```

Comment:
Hmm, so the `sizeOf (4 :: Int32)` (in this module) should be `8` right?
```bash
*Journal.Types> sizeOf (4 :: Int32)
8
```

Reply:
This is because of `type Int32 = Int` in this module, there's a comment above that type alias explaining why, basically there's no `Int32` variant of `casIntArray` (yet), once that's fixed we can remove the type alias and the size should be correct.